packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m

```diff
@@ -358,54 +360,109 @@ - (NSNumber *)getDesiredImageQuality:(NSNumber *)imageQuality {
 - (void)picker:(PHPickerViewController *)picker
     didFinishPicking:(NSArray<PHPickerResult *> *)results API_AVAILABLE(ios(14)) {
   [picker dismissViewControllerAnimated:YES completion:nil];
-
+  if (results.count == 0) {
+    self.result(nil);
+    self.result = nil;
+    _arguments = nil;
+    return;
+  }
   NSNumber *maxWidth = [_arguments objectForKey:@"maxWidth"];
   NSNumber *maxHeight = [_arguments objectForKey:@"maxHeight"];
   NSNumber *imageQuality = [_arguments objectForKey:@"imageQuality"];
   NSNumber *desiredImageQuality = [self getDesiredImageQuality:imageQuality];
 
-  for (PHPickerResult *result in results) {
+  __block NSMutableArray *pathList = [self createNSMutableArrayWithSize:results.count];
+  __block NSUInteger resultsInProgress = results.count;
+  __block dispatch_semaphore_t resultSemaphore = dispatch_semaphore_create(0);
+
+  for (int i = 0; i < results.count; i++) {
+    PHPickerResult *result = results[i];
+
     [result.itemProvider
         loadObjectOfClass:[UIImage class]
         completionHandler:^(__kindof id<NSItemProviderReading> _Nullable image,
                             NSError *_Nullable error) {
           if ([image isKindOfClass:[UIImage class]]) {
             __block UIImage *localImage = image;
-            dispatch_async(dispatch_get_main_queue(), ^{
-              PHAsset *originalAsset =
-                  [FLTImagePickerPhotoAssetUtil getAssetFromPHPickerResult:result];
+            PHAsset *originalAsset =
+                [FLTImagePickerPhotoAssetUtil getAssetFromPHPickerResult:result];
+
+            if (maxWidth != (id)[NSNull null] || maxHeight != (id)[NSNull null]) {
+              localImage = [FLTImagePickerImageUtil scaledImage:localImage
+                                                       maxWidth:maxWidth
+                                                      maxHeight:maxHeight
+                                            isMetadataAvailable:originalAsset != nil];
+            }
+            __block NSString *savedPath;
+            if (!originalAsset) {
+              // Image picked without an original asset (e.g. User pick image without permission)
+              savedPath =
+                  [FLTImagePickerPhotoAssetUtil saveImageWithPickerInfo:nil
+                                                                  image:localImage
+                                                           imageQuality:desiredImageQuality];
+              pathList[i] = savedPath;
+              if (--resultsInProgress == 0) {
+                dispatch_semaphore_signal(resultSemaphore);
+              }
```

Comment:
Several things here:
1. It seems this logic is identical as below, so it can be combined and move to the end of if-else.
2. It's not clear to me why we need to combine a semaphore with a local var. It seems redundant. 

It might be simpler to use a higher-level NSOperationQueue API for this kind of concurrency operation. Such as https://developer.apple.com/documentation/foundation/nsoperationqueue/1408358-addoperations